src/toplevel.c

```diff
@@ -585,8 +592,13 @@ jl_value_t *jl_parse_eval_all(const char *fname, size_t len)
             jl_rethrow();
         }
         else {
-            jl_rethrow_other(jl_new_struct(jl_loaderror_type, fn, ln,
-                                           jl_exception_in_transit));
+            if(jl_typeis(jl_exception_in_transit, jl_initerror_type)) {
```

Comment:
Maybe just ditch `InitError` and use `LoadError` for `__init__` errors?  After all, a `LoadError` is what you would get from the same code pasted at the end of the module.